runtime/src/test/java/com/redhat/insights/InsightsReportControllerThreadingTest.java

```diff
@@ -19,29 +19,83 @@
 
 public class InsightsReportControllerThreadingTest {
 
-  @Test
-  public void testGenerateWith403() throws Exception {
+  private static HttpResponse<String> makeResponder(int code) {
+    return new HttpResponseBase() {
+      @Override
+      public int statusCode() {
+        return code;
+      }
+    };
+  }
+
+  private static InsightsReportController makeControllerWithCode(int code) throws Exception {
     var logger = new NoopInsightsLogger(); // PrintLogger.STDOUT_LOGGER;
     var config =
         MockInsightsConfiguration.ofRetries(
             "test_app", 100L, DEFAULT_HTTP_CLIENT_RETRY_BACKOFF_FACTOR, 3);
 
     var mock = mock(HttpClient.class);
-    HttpResponse<String> response =
-        new HttpResponseBase() {
-          @Override
-          public int statusCode() {
-            return 403;
-          }
-        };
+    HttpResponse<String> response = makeResponder(code);
+
     when(mock.send(any(), any(HttpResponse.BodyHandler.class))).thenReturn(response);
     var httpClient =
         new MockInsightsJdkHttpClient(config, () -> mock(SSLContext.class), logger, mock);
 
     var report = AppTopLevelReport.of(logger);
-    var controller = InsightsReportController.of(logger, config, report, () -> httpClient);
+    return InsightsReportController.of(logger, config, report, () -> httpClient);
+  }
+
+  private static void httpCodeCausesShutdown(int code) throws Exception {
+    var controller = makeControllerWithCode(code);
     controller.generate();
 
     await().atMost(Duration.ofSeconds(5)).untilAsserted(() -> controller.isShutdown());
   }
+
+  @Test
+  public void testGenerateWith401() throws Exception {
+    httpCodeCausesShutdown(401);
+  }
+
+  @Test
+  public void testGenerateWith413() throws Exception {
+    httpCodeCausesShutdown(413);
+  }
+
+  @Test
+  public void testGenerateWith415() throws Exception {
+    httpCodeCausesShutdown(415);
+  }
+
+  @Test
+  public void testGenerateWith500() throws Exception {
+    httpCodeCausesShutdown(500);
+  }
+
+  @Test
+  public void testGenerateWith503() throws Exception {
+    httpCodeCausesShutdown(503);
+  }
+
+  @Test
+  public void testGenerateWith201() throws Exception {
+    var controller = makeControllerWithCode(201);
+
+    Runnable r = () -> controller.generate();
+    Thread t = new Thread(r);
+    t.start();
+    t.join();
```

Comment:
Just to be sure: you want to check that the shutdown happens within 5 seconds after the generation has finished, right?

Reply:
Yes, and that the controller will actually exit after the spawned thread (the last non-daemon thread) has completed